### ISSUE TITLE
Custom elements duplicated

### ISSUE THREAD

**Issue by crazytyper** [PDFTron/webviewer-ui#345]
If function `render()` of a header item of type `customElement` returns a DOM element, then the content of this custom element is duplicated every time `setHeaderItems()` is called.

![duplicate_custom_elements](https://user-images.githubusercontent.com/1025263/63936641-96f7bc80-ca60-11e9-88fc-0472bf7b6874.png)

The reason for this is pretty obvious:
https://github.com/PDFTron/webviewer-ui/blob/cc623f2b04f6fbe62b7d9b56e9272a1530f0e2c9/src/components/CustomElement/CustomElement.js#L48-L52

IMO the existing children have to be removed before appending the new DOM element.

**Comment by crazytyper** [comment]
I have updated `App.js` from https://github.com/PDFTron/webviewer-react-sample to reproduce the issue:

```jsx
import React, { Component } from "react";
import "./App.css";

const buildButtons = () => {
    return [
        {
            type: "customElement",
            render() {
                const el = document.createElement("h1");
                el.textContent = "😁";
                return el;
            }
        }
    ];
};

class App extends Component {
    constructor() {
        super();
        this.viewer = React.createRef();
    }

    componentDidMount() {
        window.WebViewer({ path: "/lib" }, this.viewer.current).then(instance => {
            this.instance = instance;

            this.instance.setHeaderItems(header => {
                header.update(buildButtons());
            });
            this.instance.setHeaderItems(header => {
                header.update(buildButtons());
            });
            this.instance.setHeaderItems(header => {
                header.update(buildButtons());
            });
        });
    }

    render() {
        return (
            <div className="App">
                <div className="header">React sample</div>
                <div className="webviewer" ref={this.viewer}></div>
            </div>
        );
    }
}

export default App;
```

**Comment by henry-chris** [comment]
I have also had this issue for awhile.  I just thought it was me messing something up.  For me, it only happened during development (with hotreload active) but not in production.

Once we upgraded to 5.2+ and added ReactHotLoader the problem seems to have gone away.  However, the development environment seems very very slow now.  Much slower loads and the Redux Dev Logger seems to just crawl (meaning messages are output about 1 for every 0.5 s).

Not sure if that's the current HMR status or something that happened with the new React hooks makeover.

**Comment by ZhijieZhang** [comment]
@crazytyper Thanks. I'm able to reproduce the issue on my end. I think what's happening here is that the reference of the `render` function of the button object returned by `buildButtons` is always new so the `useEffect` will run every time `header.update` is called.

I think there're two options to resolve this issue:
1. As you suggested, remove the children in the effect before appending the new one. So the code will become something like:
```javascript
if (isDOMElement(element)) {
  const wrapperElement = wrapperRef.current;

  while (wrapperElement.firstChild) {
    wrapperElement.removeChild(wrapperElement.firstChild);
  }
  wrapperElement.appendChild(element);
}
```
2. Don't run this effect when the `props.render` changes. This is actually how the component behaves when it was written using class instead of hooks. However I don't think this is a proper solution the more I think about it. I think I will go with option 1 and include the fix in 5.2.4. I will give an update after it's available.

Again, thanks for bring this up!

**Comment by ZhijieZhang** [comment]
@henry-chris Yeah I noticed this as well. The UI is running slowly in the dev build, especially when the console is open and printing out the states from redux-logger. Not sure if I'm missing something configurations or other things aren't going well.

**Comment by ZhijieZhang** [comment]
Hi @crazytyper,

We just released the 6.0-beta, which can be downloaded from https://www.pdftron.com/downloads/WebViewer-6.0.0-beta-1.zip.
We also have a migration guide here https://www.pdftron.com/documentation/web/get-started/migrating-to-v6.

I believe the issue you were seeing is fixed in this build. Can you give it a try and let us know how it goes? Thanks!

**Comment by ZhijieZhang** [comment]
Hi, I will close the ticket for now. Please feel free to reopen it if you found any issues in the 6.0-beta build.
